### ISSUE TITLE
About prot_t5

### ISSUE THREAD

**Issue by yelou2022** [agemagician/ProtTrans#154]
Hello, thank you very much for your work. I have a few small doubts to ask:

1. I found in the Embedding folder that when using T5 to generate the embedding representation of the sequence, the model will be set to evaluation mode, but it is not set in the sample code of the README.md file. So if T5 is used as an encoder to extract the embedding representation of the sequence, do I need to set this?

2. In hugging face, I found that T5 has two versions, prot_t5_xl_uniref50 and prot_t5_xl_half_uniref50-enc. Is there any difference between them, or is it just the difference between full-precision and half-precision? Because I found that the sizes of the two models on the website are different, one is 11.3G and the other is 2.42G. Does it mean that prot_t5_xl_uniref50 is better than prot_t5_xl_half_uniref50-enc?

Thanks for the reply and wish you a happy life!

**Comment by mheinzinger** [comment]
Hi, sorry for the delayed response and thanks for your interest in our work :) 
1. Good catch; if you want to extract embeddings, you want reproducible results which is why the model should be run in evaluation mode (de-activating e.g. dropout). Making this explicit was for me rather a sanity check to be really sure it is set correctly but I do not think its strictly needed as the [.from_pretrained()](https://huggingface.co/docs/transformers/main_classes/model#transformers.PreTrainedModel.from_pretrained) method sets the model to evaluation mode by default.
2. The difference is that prot_t5_xl_uniref50 holds the full ProtT5 model (encoder+decoder) in full-precision. Whereas prot_t5_xl_half_uniref50-enc only holds the encoder-part of ProtT5 in half-precision (therefor, you see the reduction of size by a factor of 4 - cutting by half when stripping off the decoder and cutting again by half when going from fp32 to fp16). If you are only interested in extracting embeddings, you are good2go with the fp16, encoder-only version. But if your use-case requires running the decoder (generation, finetuning, log-odds extraction etc.), then you need the fp32 enc.-dec. version. So its not a matter of "better" bc the encoder-weights are always the same but rather depends on your specific use-case.

**Comment by yelou2022** [comment]
> Hi, sorry for the delayed response and thanks for your interest in our work :)
> 
> 1. Good catch; if you want to extract embeddings, you want reproducible results which is why the model should be run in evaluation mode (de-activating e.g. dropout). Making this explicit was for me rather a sanity check to be really sure it is set correctly but I do not think its strictly needed as the [.from_pretrained()](https://huggingface.co/docs/transformers/main_classes/model#transformers.PreTrainedModel.from_pretrained) method sets the model to evaluation mode by default.
> 2. The difference is that prot_t5_xl_uniref50 holds the full ProtT5 model (encoder+decoder) in full-precision. Whereas prot_t5_xl_half_uniref50-enc only holds the encoder-part of ProtT5 in half-precision (therefor, you see the reduction of size by a factor of 4 - cutting by half when stripping off the decoder and cutting again by half when going from fp32 to fp16). If you are only interested in extracting embeddings, you are good2go with the fp16, encoder-only version. But if your use-case requires running the decoder (generation, finetuning, log-odds extraction etc.), then you need the fp32 enc.-dec. version. So its not a matter of "better" bc the encoder-weights are always the same but rather depends on your specific use-case.

Thanks, wish you a good day!
